### PR TITLE
feat(target_chains/ton): add execute_governance_action

### PR DESCRIPTION
**Pull request by cctdaniel**

### CODE REVIEW

File: target_chains/ton/contracts/contracts/Pyth.fc [hunk 1]
```diff
@@ -203,6 +220,134 @@ int parse_pyth_payload_in_wormhole_vm(slice payload) impure {
...
+    int result = value;
+    repeat (expo) {
+        result *= 10;
+    }
+    return result;
+}
+
+() execute_upgrade_contract(slice payload) impure {
     ;; TODO: Implement
 }
```
Comment (cctdaniel) [inline target_chains/ton/contracts/contracts/Pyth.fc:255]: will implement in a following PR

File: target_chains/ton/contracts/tests/utils/pyth.ts [hunk 2]
```diff
@@ -36,3 +36,26 @@ export const HERMES_BTC_ETH_UPDATE =
 
 export const BTC_PRICE_FEED_ID =
   "0xe62df6c8b4a85fe1a67db44dc12de5db330f7ac66b72dc658afedf0f4a415b43";
+
+export const TEST_GUARDIAN_ADDRESS1 =
+  "0x686b9ea8e3237110eaaba1f1b7467559a3273819";
+
+// A Pyth governance instruction to authorize governance data source transfer signed by the test guardian #1.
```
Comment (ali-bahjati) [inline target_chains/ton/contracts/tests/utils/pyth.ts:43]: not for this PR but for future PRs: this copying is brittle. We should at some point (during the development of TON) elevate the rust test generating code from starknet and change it to generate this file.

File: target_chains/ton/contracts/tests/PythTest.spec.ts [hunk 3]
```diff
@@ -282,4 +305,198 @@ describe("PythTest", () => {
 
     // TODO: add more tests for other governance data source
   });
+
```
Comment (ali-bahjati) [inline target_chains/ton/contracts/tests/PythTest.spec.ts:308]: i see a lot of happy paths here, can you add some tests for the failure cases? (Same for price updates)

**Review by ali-bahjati — APPROVED** [review]
Nice! I'm approving this but we need to add more test cases for failure scenarios and i strongly recommend changing the rust test case generator to generate typescript fixtures instead of using its outputs in cairo. Having a generic test case generator will massively help us in new network integrations.

File: target_chains/ton/contracts/contracts/common/gas.fc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+int get_compute_fee(int workchain, int gas_used) asm(gas_used workchain) "GETGASFEE";
+
+;; The actual gas used for the transaction is 350166 but we add ~10% (385182.6) and round up (390000) to be on the safe side because the amount of gas used can vary based on the current state of the blockchain
+const int UPDATE_PRICE_FEEDS_GAS = 390000;
```
Comment (cctdaniel) [inline target_chains/ton/contracts/contracts/common/gas.fc:4]: calculated by following the best practice suggested [here](https://docs.ton.org/develop/smart-contracts/fee-calculation#calculation-flow-1) which is to run tests and get the actual `gas_used` amount

File: target_chains/ton/contracts/contracts/common/constants.fc [hunk 5]
```diff
@@ -9,3 +9,5 @@ const int GUARDIAN_SET_EXPIRY = 86400; ;; 1 day in seconds
 const int UPGRADE_MODULE = 0x0000000000000000000000000000000000000000000000000000000000436f7265; ;; "Core" (left-padded to 256 bits) in hex
 
 const int WORMHOLE_MERKLE_UPDATE_TYPE = 0;
+
+const int WORKCHAIN = 0;
```
Comment (ali-bahjati) [inline target_chains/ton/contracts/contracts/common/constants.fc:13]: What is this?

**Review by ali-bahjati — APPROVED** [review]

File: target_chains/ton/contracts/contracts/common/constants.fc (hunk 5 above)
Comment (cctdaniel) [reply target_chains/ton/contracts/contracts/common/constants.fc:13]: there are 2 blockchains on TON: https://docs.ton.org/learn/overviews/ton-blockchain#blockchain
and they have different costs depending on the chains: https://docs.ton.org/learn/overviews/ton-blockchain#blockchain

will add comment to the file to make it clearer

**Pull request merged**